PEP 772: Packaging governance process

pradyunsg opened the pull request:
<!--
You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.

If your PEP is not Standards Track, remove the corresponding section.
-->

## Basic requirements (all PEP Types)

* [x] Read and followed [PEP 1](https://peps.python.org/1) & [PEP 12](https://peps.python.org/12)
* [x] File created from the [latest PEP template](https://github.com/python/peps/blob/main/peps/pep-0012/pep-NNNN.rst?plain=1)
* [x] PEP has next available number, & set in filename (``pep-NNNN.rst``), PR title (``PEP 123: <Title of PEP>``) and ``PEP`` header
* [x] Title clearly, accurately and concisely describes the content in 79 characters or less
* [x] Core dev/PEP editor listed as ``Author`` or ``Sponsor``, and formally confirmed their approval
* [x] ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
* [x] ``PEP-Delegate``, ``Topic``, ``Requires`` and ``Replaces`` headers completed if appropriate
* [x] Required sections included
    * [x] Abstract (first section)
    * [x] Copyright (last section; exact wording from template required)
* [x] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer names](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
* [x] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
* [x] Authors/sponsor added to ``.github/CODEOWNERS`` for the PEP


## Standards Track requirements

* [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate
* [x] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [x] Motivation
    * [x] Rationale
    * [x] Specification
    * [ ] Backwards Compatibility
    * [ ] Security Implications
    * [ ] How to Teach This
    * [ ] Reference Implementation
    * [x] Rejected Ideas
    * [ ] Open Issues
* [x] ``Python-Version`` set to valid (pre-beta) future Python version, if relevant
* [x] Any project stated in the PEP as supporting/endorsing/benefiting from the PEP formally confirmed such
* [x] Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in ``Discussions-To`` and ``Post-History``


<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--4218.org.readthedocs.build/pep-0772/

<!-- readthedocs-preview pep-previews end -->

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Status: Draft
+Type: Process
+Topic: Governance, Packaging
+Created: 31-Dec-2024
```
This is the date the PEP number is assigned:
```suggestion
Created: 21-Jan-2025
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Abstract
+========
+
+This PEP proposes a Python Packaging Council with broad authority over packaging standards, tools, and implementations. Like the Python Steering Council, they seek to exercise this authority as rarely as possible; instead, they use this power to establish standard processes.
```
Please wrap all lines to ~79 chars:

https://peps.python.org/pep-0012/#general

Tip: you can use something like `prettier --prose-wrap=always --print-width=88 --write peps/pep-0772.rst`

pradyunsg replied:
FWIW:

```
❯ npx prettier --prose-wrap=always --print-width=88 --write peps/pep-0772.rst
Need to install the following packages:
prettier@3.4.2
Ok to proceed? (y) y

[error] No parser could be inferred for file "/Users/pradyunsg/Developer/github/peps/peps/pep-0772.rst".
[[[I trimmed the rest of the noisy error]]]
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Approval voting in the election
+-------------------------------
+
+An earlier non-public draft of this PEP utilised an approval voting process, which aligned with what :pep:`13` stated at the time of writing. The Python core team has changed their governance to utilise BLOC-STAR and this PEP was updated to align with that for the same reasons as the core teams' move to BLOC STAR: it better captures voter intention in the results.
```
```suggestion
An earlier non-public draft of this PEP used an approval voting process, which aligned with what :pep:`13` stated at the time of writing. The Python core team has changed their governance to use Bloc STAR and this PEP was updated to align with that for the same reasons as the core team's move to Bloc STAR: it better captures voter intention in the results.
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Packaging-WG
+------------
+
+The PSF's `Packaging Workgroup`_ was created to support the larger efforts of improving and maintaining the packaging ecosystem in Python through fundraising and disbursement of raised funds. It was expected to largely focus on efforts such as PyPI, pip, packaging.python.org, setuptools, and cross-project efforts.
```
```suggestion
The PSF's `Packaging Workgroup`_ was created to support the larger efforts of improving and maintaining the packaging ecosystem in Python through fundraising and disbursement of raised funds. It was expected to largely focus on efforts such as PyPI, pip, packaging.python.org, Setuptools, and cross-project efforts.
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Packaging Council
+-----------------
+
+The Packaging Council will be composed of 5 individuals.
```
Maybe?
```suggestion
The Packaging Council will be composed of 5 people.
```

pradyunsg replied:
There's multiple spots using this phrasing, so I'd prefer to leave this as-is.

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+
+The Python Steering Council would be expected to delegate decision making to the Packaging Council for PEPs related to the Python packaging.  The bodies would work together on issues that intersect the packaging domain and language stewardship (including the CPython implementation, standard library, and distribution).
+
+The PSF Board would be encouraged to formally deprecate the Packaging workgroup and the Packaging Council would wholly take on the responsibilities of the PSF's Packaging workgroup.
```
Caps as above:
```suggestion
The PSF Board would be encouraged to formally deprecate the Packaging Workgroup and the Packaging Council would wholly take on the responsibilities of the PSF's Packaging Workgroup.
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Conflicts of interest
+---------------------
+
+No more than two Packaging Council members should be employed by or significantly affiliated with the same entity.
```
What is an entity? A company? Or can it be something else?

pradyunsg replied:
Clarified in a follow-up change.

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+* PyPA members: Anyone with the triage bit or commit bit, or at least one project in the PyPA organisation.
+* Packaging workgroup members: Anyone who is listed on the Packaging WG charter will be moved into the Packaging Community.
+* Interested core team members: Any Python core team member who is willing to participate is welcome.
+* Wider community members: Non-profit organisations that participate in packaging or working with new packagers. For example, PyOpenSci, NumFocus, Django, are encouraged to initially nominate up to 7 members by sending an email to \[todo\].
```
```suggestion
* Wider community members: non-profit organisations that participate in packaging or working with new packagers. For example, PyOpenSci, NumFocus, Django, are encouraged to initially nominate up to 7 members by sending an email to \[todo\].
```

pradyunsg replied:
The `:` is acting as a separator and all other entries in this list use a capitalized word after it. I'd prefer to keep the current style here.

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+
+Members are added to the Packaging community by a simple majority vote by the current membership. Quorum for adding new members is 50%.
+
+A vote to add a new member is triggered when a Packaging community member calls for one publicly on an appropriate communication channel, and another Packaging community member seconds the call within two weeks.
```
How long is the actual vote open for?

Perhaps one week, like for the core team? https://peps.python.org/pep-0013/#membership

pradyunsg replied:
Good point! Not specifying this was an oversight.

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+
+In order to maintain a reasonable expectation of quorum, failure to participate in Packaging Council elections for two consecutive council elections automatically removes a person from the list of voting members, until they re-submit their intention to resume their participation to the Packaging Council in writing.
+
+In exceptional circumstances, it may be necessary to remove someone from the Packaging community against their will (for example: egregious and ongoing code of conduct violations). A Packaging community member may be removed by a two-thirds majority vote by the Packaging Council (in practice: 4:1 for a council with 5 individuals). If the ejected Packaging community member is also on the Packaging ...
```
```suggestion
In exceptional circumstances, it may be necessary to remove someone from the Packaging community against their will (for example: egregious and ongoing code of conduct violations). A Packaging community member may be removed by a two-thirds majority vote by the Packaging Council (in practice: 4:1 for a council with 5 members). If the ejected Packaging community member is also on the Packaging Council, then they are removed from the council as well.
```

hugovk replied:
What happens if this person is on the Packaging Council, do they get to participate in the vote?

pradyunsg replied:
> What happens if this person is on the Packaging Council, do they get to participate in the vote?

Yes, and it's expected that they'd vote "no" and still get ejected FWIW.

pradyunsg replied:
I'll clarify the language.

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+Annual elections for all council members
+----------------------------------------
+
+An annual term for council members is the approach taken for the Python steering council's elections. This PEP uses a cohort-based model, derived from the PSF Board's elections which enables continuity of members across a changing council.
```
Same caps as above:
```suggestion
An annual term for council members is the approach taken for the Python Steering Council's elections. This PEP uses a cohort-based model, derived from the PSF Board's elections which enables continuity of members across a changing council.
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+
+This PEP currently mirrors the Python Steering Council's limit, that at most two individuals related to a single organisation can be on the council.
+
+Limiting it to one is workable; although it hasn't come up in the SC, people do move around, and we wouldn't want good candidates to either make employment decisions based on PC membership, or have to resign based on an employment change.  Limiting it to max 2, plus votes of no confidence is probably sufficient safety from any undo employer influence.
```
```suggestion
Limiting it to one is workable; although it hasn't come up in the SC, people do move around, and we wouldn't want good candidates to either make employment decisions based on PC membership, or have to resign based on an employment change.  Limiting it to max 2, plus votes of no confidence is probably sufficient safety from any undue employer influence.
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,292 @@
+
+To that end, the process for approval for this PEP will be:
+
+* Submit this PEP for a vote on pypa-committers, in accordance with the process outline in :pep:`609`.
```
```suggestion
* Submit this PEP for a vote on the pypa-committers mailing list, in accordance with the process outlined in :pep:`609`.
```

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+decision making process that governs the technical projects under the PyPA
+umbrella.
+
+Processes
+=========
```
The process set out in this section seems very "heavy", for want of a better word. Could you add some text to Rejected Ideas for why a [PEP 729](https://peps.python.org/pep-0729/#operations-and-process)-like approach wasn't adopted?

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+
+Each phase should last two weeks.
+
+The election process is managed by a returns officer nominated by the outgoing
+Packaging Council. For the initial election, the returns officer will be
+nominated by the PSF Executive Director.
```
```suggestion
The election process is managed by a Returning Officer nominated by the outgoing
Packaging Council. For the initial election, the Returning Officer will be
nominated by the PSF Executive Director.
```

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+
+* Phase 1: A member of the Packaging community (defined later in this document)
+  can self-nominate themselves for the council elections.
+* Phase 2: Each core team member can assign zero to five stars to each
+  candidate. Voting is performed anonymously. The outcome of the vote is
+  determined using the `STAR voting system`_, modified to use the `Multi-winner
+  Bloc STAR`_ approach. If a tie occurs, it may be resolved by mutual agreement
+  among the candidates, or else the winner will be chosen at random.
```
Is this "core team" the core developers/committers? If not, could you disambiguate? If so, it seems a bit surprising--would it be better to have the packaging community members form the electorate?

pradyunsg replied:
Whoops! That's a copy-paste mistake.

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+Term
+----
+
+There shall be two cohorts of council members: Cohort A composed of two (2)
+members and Cohort B composed of three (3) members.
+
+Each council member shall be elected for a two-year term, unless they are
+replacing a council member that resigned, was removed, or otherwise becomes
+vacant, in which case such replacement members shall be elected to a term
+sufficient to complete a two-year term as measured from the term of the
+original cohort.
+
+Each council member's term will be two (2) years, from when their elections are
+finalised until the next elections for their cohort are finalised.
+
+Only for elections involving the entire council (like the initial council
+election), the two candidates receiving the highest number of votes shall be
+designated Cohort A with a two (2) year term, and the three candidates
+receiving the highest number of votes after shall be designated Cohort B with a
+one (1) year term.
+
+There are no term limits for individual council members.
```
As a proceduralist myself, I hesistate to say that this seems overly formalistic -- experience with the steering council shows a lot of natural overlap. If you want to stagger the terms, perhaps just say:

```
For the first election of the Council, and any subsequent elections where
all five positions are vacant, the two highest-ranking candidates shall be
elected for a two-year term, and the next three highest-ranking candidates
shall be elected for a one-year term.
```

pradyunsg replied:
Is this suggestion to replace all the text? Because it's missing the wording around the two year term stuff.

FWIW, this text is modelled after the PSF's bylaws, so... yea, it's definitely very formal. 😅

warsaw replied:
@AA-Turner the only scenario I can think of where all five positions are vacant _except_ for the first vote would be a mass resignation of the PC.  I think the original text covers that; are you thinking of something else?

FWIW, the two-cohort idea has been floated for the SC, to ensure continuity, but a) it fortunately hasn't come up in practice yet; b) would require a PEP 13 change process to achieve.  This is one of those sections that deliberately deviate from PEP 13 rhyming, I think for a good reason.

pradyunsg replied:
Re-reading the first comment and @warsaw's comment today... I feel like this is a suggestion for changing about what's being proposed, rather than something editorial about the way things are written.

If my understanding is correct, I suggest we defer further discussion on this until the PEP is up on discuss.python.org (to avoid non-editorial discussions in this PR).

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+
+Council members may resign their position at any time.
+
+Whenever there is a vacancy during the regular council term, the council may
+vote to appoint a replacement to serve out the rest of the term.
```
This seems contrary to the above section, which regulates for by-elections. Is having both co-option and by-elections intentional?

```suggestion
Whenever there is a vacancy during the regular council term, the council may
vote to appoint a replacement to serve the remainder of the vacant term.
```

pradyunsg replied:
I don't understand the question being asked here. 🙈 

As I understand this, "rest of the term" here is described in the "Term" section before this:

> Each council member shall be elected for a two-year term, unless they are
replacing a council member that resigned, was removed, or otherwise becomes
vacant, in which case such replacement members shall be elected to a term
sufficient to complete a two-year term as measured from the term of the
original cohort.

If the current language isn't clear, do you think the following language would be clearer here?

```suggestion
Whenever there is a vacancy during the regular council term, the council may
vote to appoint a replacement.
```

warsaw replied:
I'm going to resolve this thread without change.  I feel like the original language is clear enough, but if clarification is necessary, we can discuss this on the upcoming DPO thread.

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+Whenever there is a vacancy during the regular council term, the council may
+vote to appoint a replacement to serve out the rest of the term.
+
+If a council member drops out of touch and cannot be contacted for a month or
+longer, then the rest of the council may vote to replace them (with a simple
+majority vote, not counting the relevant member).
```
The parenthetical aside is covered by both "the rest of the council" and that there is no supermajority voting specified above.

This section also suggests co-option rather than by-election.
```suggestion
If a council member cannot be contacted for a month or longer,
then the rest of the council may vote to replace them.
```

pradyunsg replied:
I'm gonna take the liberty of marking this as resolved, since it looks like we're gonna stick with the current language here.

Please feel welcome to click the unresolve button on this if this language is something you think should change!

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+---------------------
+
+No more than two Packaging Council members should be employed by or
+significantly affiliated with the same entity.
```
What does "entity" mean here? The PSF? A particular packaging project or tool (e.g. pip)?

pradyunsg replied:
This bit has been re-written to be clearer -- thanks for flagging this!

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+6th-ranking candidate moves up into 5th place; this is repeated until a valid
+council is formed.
+
+During a council term, if changing circumstances cause this rule to be broken
+(for instance, due to a council member changing employment), then one or more
+council members must resign to remedy the issue, and the resulting vacancies
+can then be filled as normal.
```
How is the resign-ee chosen here?

willingc replied:
I think that it is specified above. The remaining council may choose to appoint a replacement member. The composition of the council should still maintain a max of 2 individuals from the same company, etc.

pradyunsg replied:
It is specified above. I'm gonna take the liberty of marking this resolved without making any text changes here.

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+  will be moved into the Packaging Community.
+* Interested core team members: Any Python core team member who is willing to
+  participate is welcome.
+* Wider community members: Non-profit organisations that participate in
+  packaging or working with new packagers. For example, PyOpenSci, NumFocus,
+  Django, are encouraged to initially nominate up to 7 members by sending an
+  email to \[todo\].
```
Seven each or total? Why 7?

willingc replied:
The limit or goal of 7 seems a bit arbitrary.

warsaw replied:
> The limit or goal of 7 seems a bit arbitrary.

It kinda is!  Do you think there shouldn't be any limit or a different limit?  I think the intent was so that no one org could overwhelm the pool of members.

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+ecosystem-wide changes are being proposed. In particular, the work to support
+existing projects and maintain a user guide has been excellent.
+
+At the time of writing, PyPA is still less formal than its counterpart on the
+CPython side. It does not have a community elected body and individuals within
```
What is the "counterpart" referring to here?

pradyunsg replied:
PyPA <-> CPython core + steering council, essentially.

pradyunsg replied:
> remove reference here to CPython

Did this here.

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+Python Steering Council
+-----------------------
+
+While there is overlap between the PyPA, Packaging-WG as well as the Python
+core team, the Steering Council is not well-placed to make decisions over
```
```suggestion
While there is overlap between the PyPA, the Packaging-WG, and the Python
core team, the Steering Council is not well-placed to make decisions over
```

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+An elected Packaging Council that has authority over the interoperability
+standards and Python packaging tooling would have the mandate to coordinate
+efforts around Python packaging. This would provide a better, more consistent
+experience to packagers, package consumers, tool developers, PyPI, and other
```
I think this is the first time PyPI is used?
```suggestion
experience to packagers, package consumers, tool developers, the Python Package Index (PyPI), and other
```

pradyunsg replied:
It's earlier at line 64, while I'll update to call out this acronym like this!

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+The council shall:
+
+* have broad authority over the Python packaging standards and Python Packaging
+  User Guide, that are maintained on packaging.python.org.
```
```suggestion
  User Guide, that are maintained on https://packaging.python.org/.
```

AA-Turner commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+* establish processes for making binding decisions regarding packaging
+  standards, tools and implementations as well as for considering
+  ecosystem-wide changes.
+* look for ways to use their direct authority as little as possible – seeking
+  consensus/consent over voting.
```
```suggestion
* look for ways to use their direct authority as little as possible – seeking
  consensus and consent over voting.
```

zanieb commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+The council shall:
+
+* have broad authority over the Python packaging standards and Python Packaging
```
Nit: elsewhere you capitalize list items

zanieb commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+  will be moved into the Packaging Community.
+* Interested core team members: Any Python core team member who is willing to
+  participate is welcome.
+* Wider community members: Non-profit organisations that participate in
```
Does this exclude for-profit organizations? If so, should that be mentioned in detail somewhere?

willingc replied:
Good catch @zanieb. It should not exclude for-profit members working on an open source project IMHO.

pradyunsg replied:
We should clear this up before the initial d.p.o thread.

I'll talk out-of-band with the co-authors and update this PEP when we have better language around this.

warsaw replied:
I think it's probably highly likely that any candidate for initial membership will have *some* affiliation to non-profit or OSS orgs, in addition to any for-profit org they may be affiliated with.  We can discuss on the DPO thread, but I would be surprised if there really is anybody excluded from membership because of this.

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Conflicts of interest
+---------------------
+
+No more than two Packaging Council members should be employed by or
+significantly affiliated with the same entity. An entity would be a company,
+a company and its subsidiaries, or another incorporated entity such as a
+non-profit or educational institution with its own mission and goals.
```
About non-profits, I can see you may want to limit to, say, two DSF members, but I imagine it would be impractical and not your aim to limit to two PSF members 🙃 

Perhaps mention the PSF as excluded?

willingc replied:
I would say on a council of 5 that 2 PSF employees would be a maximum on the council.

willingc replied:
I would be explicit about PSF members vs. PSF employees since I would expect most candidates to be PSF members.

warsaw replied:
PEP 13 says: "In order to avoid any appearance of conflict of interest, at most 2 members of the council can work for any single employer" so I don't think the same ambiguity exists for the SC.

pradyunsg replied:
Yea, this bit here isn't derived from other governance documents... We should clear this up before the initial d.p.o thread.

I'll talk out-of-band with the co-authors and update this when we have an answer.

warsaw replied:
```suggestion
No more than two Packaging Council members should be employed by or
significantly affiliated with the same entity. An entity would be a company,
a company and its subsidiaries, or another incorporated entity such as a
non-profit or educational institution with its own mission and goals.  PSF members are explicitly exempt from this restriction, but _not_ PSF employees, directors, or officers.
```

hugovk commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Type: Process
+Topic: Governance, Packaging
+Created: 21-Jan-2025
+Replaces: 609
```
PEP editors: should we add `Superseded-By: 772` to PEP 609, or only when this one is accepted?

pradyunsg replied:
(not an editor, but author on both)

I'd prefer that we only add it once this PEP is accepted.

warsaw commented:
I thought we would put this in PEP 8800 (or other 8X00 range) to rhyme with the Steering Council PEPs in the 8100s?

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Expectations
+------------
+
+An elected Packaging Council that has authority over the interoperability
+standards and Python packaging tooling would have the mandate to coordinate
+efforts around Python packaging. This would provide a better, more consistent
+experience to packagers, package consumers, tool developers, PyPI, and other
```
```suggestion
An elected Packaging Council would have authority over the interoperability
standards, would guide the packaging tooling used by CPython, and have the mandate to coordinate Python packaging efforts. This would provide a better, more consistent
experience to packagers, package consumers, tool developers, PyPI, and other
```

pradyunsg replied:
> would guide the packaging tooling used by CPython

I think this would be a change of scope (we probably don't want the Packaging council involved in things like PEP 773, which this language would imply.

I'm incorporating this suggestion, with a minor rephrasing as "would guide the packaging tooling provided with CPython".

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+The council shall work to:
+
+* Improve Python packaging's user experience.
+* Maintain the quality and stability of the Python packaging standards.
+* Formalise and maintain the relationship with the core team as well as the
+  PSF.
+* Establish appropriate decision-making processes.
+* Seek consensus among contributors before acting in a formal capacity.
+* Make contributing as accessible, inclusive, and sustainable as possible.
```
I suggest that you order these points in priority of short term and long term efforts. Perhaps: 
```suggestion
* Maintain the quality and stability of the Python packaging standards.
* Formalise and maintain the relationship with the core team as well as the
  PSF.
* Establish appropriate decision-making processes.
* Improve Python packaging's user experience.
* Make contributing as accessible, inclusive, and sustainable as possible.
* Strive to seek consensus among contributors before acting in a formal capacity.
```

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+To use its powers, the council votes. Every council member must either vote or
+explicitly abstain. Members with conflicts of interest on a particular vote
+must abstain. Passing requires support from a majority of non-abstaining
```
I would recommend that a minimum quorum of 3 voting members should be used when there are multiple abstentions. If more than 2 members abstain, that begs the question is the item up for vote well scoped for decision making.

willingc replied:
Good question @warsaw. In general, it shouldn't be a problem. My concern is that 4 people abstain and the decision falls on 1 person's shoulders. I'm fine with mirroring PEP 13 unless necessary to diverge.

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+distribution).
+
+The PSF Board would be encouraged to formally deprecate the Packaging Workgroup
+and the Packaging Council would wholly take on the responsibilities of the
```
```suggestion
and the Packaging Council would take on the responsibilities of the
```

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+and the Packaging Council would wholly take on the responsibilities of the
+PSF's Packaging Workgroup.
+
+The PyPA would be expected to work with the Packaging Council to establish a
```
This feels ambiguous to me. I would view the Packaging Council has broad authority over the technical projects under the PyPA umbrella on standards and strategic fit in the packaging ecosystem. I would expect the individual projects under the PyPA to make day to day decisions. As written, the PyPA and Packaging Council seem equal in decision making with no clear authority over strategic direction/evolution of packaging.

pradyunsg replied:
This is marked as outdated by GitHub because the language changed slightly:

> The PyPA _is_ expected to work with the Packaging Council to establish a
decision making process that governs the technical projects under the PyPA
umbrella.

The concern raised is still valid.


(at least from my perspective) This ambiguity is intentional currently. The specifics of "what the PC <-> PyPA relationship should be" is something that we want the inaugral council to figure out, rather than front-loading & coupling that discussion with the establishment of a council that (at minimum) solves the PEP delegate bottleneck issue.

I expect this to be a major point of discussion when we create a d.p.o thread for this PEP. I also expect that there are gonna be a lot of opinions around how this should work and a choice will need to be that'll make someone/everyone slightly/very unhappy -- and I'd like for the inaugural council to be making that choice, with ideally the candidates being elected describing how they expect the relationship to work.

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+* Phase 1: A member of the Packaging community (defined later in this document)
+  can self-nominate themselves for the council elections.
+* Phase 2: Each core team member can assign zero to five stars to each
```
"Core team member" needs more explicit definition or reference to its definition. Is there a published list of core team members now?

hugovk replied:
I think the private https://github.com/python/voters repo has the definitive list (119 members). There's also the public https://devguide.python.org/core-developers/developer-log/ (119 members) which might not quite align, and https://github.com/orgs/python/teams/python-core/members (131 members) which I'm not sure is public.

The annual SC election PEPs at https://peps.python.org/topic/governance/ list the active team members for each election (last time there were 100 https://peps.python.org/pep-8106/#complete-voter-roll) and is generated from https://github.com/python/voters.

hugovk replied:
Actually, I think "core team" here is referring to "PyPA community membership", and this is what @willingc says needs clarifying (and as my confusion shows!)

We can reference the PEP 13 section further down in:

> * Interested core team members: Any Python core team member who is willing to
  participate is welcome.

willingc replied:
Yes, I would like clarification of Packaging core team. I'm not sure who that includes.

pradyunsg replied:
FWIW, this was a copy-paste mistake -- this has now been amended to correctly state "Packaging community member" instead of "core team member".

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+Each phase should last two weeks.
+
+The election process is managed by a returns officer nominated by the outgoing
```
```suggestion
The election process is managed by an elections officer nominated by the outgoing
```

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Each phase should last two weeks.
+
+The election process is managed by a returns officer nominated by the outgoing
+Packaging Council. For the initial election, the returns officer will be
```
```suggestion
Packaging Council. For the initial election, the elections officer will be
```

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+The election process is managed by a returns officer nominated by the outgoing
+Packaging Council. For the initial election, the returns officer will be
+nominated by the PSF Executive Director.
```
```suggestion
selected by the PSF Executive Director.
```

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+If a council member drops out of touch and cannot be contacted for a month or
+longer, then the rest of the council may vote to replace them (with a simple
+majority vote, not counting the relevant member).
```
```suggestion
majority vote, where the missing member is recorded as an abstention).
```

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Responsibility
+==============
+
+Packaging community members participate in formal votes to elect the Packaging
```
This seems to conflict with the earlier mention of Packaging core.

pradyunsg replied:
That mention has been corrected to say "Packaging community members".

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+
+To that end, the process for approval for this PEP will be:
+
+* Submit this PEP for a vote on the pypa-committers mailing list, in accordance
```
The list of individuals on pypa-committers should be added as an appendix to this PEP.

pradyunsg replied:
I think that mailing list's membership isn't strictly accurate today, and this is a detail we should clean this up before the actual vote.

I've made a note of this on my end, as a thing to follow up on separately -- there might be a case for adding an open question entry for the specifics of how to determine who is "actually" supposed to be a part of pypa-committers.

willingc commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+* The Packaging Council (similar to the Steering Council) is encouraged to
+  communicate with and when necessary seek advice from the PSF's Conduct
+  Working Group.
+* Find ways to regularly synchronise with the Steering Council.
```
Perhaps initially a quarterly joint meeting.

warsaw replied:
I think the SC and PC could work out the cadence, much like the SC does with the DiR syncs.

willingc replied:
Maybe recommend an initial cadence, subject to change by SC and PC,  so the expectation isn't weekly or monthly.

warsaw replied:
How about "Regularly synchronize with the Steering Council on a mutually agreed cadence, with a recommended frequency of no less than once per quarter."

willingc replied:
@warsaw Perfect! TYVM.

willingc reviewed (commented):
Overall, nice work on this PEP. I'm happy to see it moving forward. I've made some comments based on my governance experience. YMMV.

hugovk commented:
> I thought we would put this in PEP 8800 (or other 8X00 range) to rhyme with the Steering Council PEPs in the 8100s?

Hmm, no objection, although this is replacing 609, and the other recent council/WG/board PEPs didn't go in 8xxx:

  | PEP | Title | Authors
-- | -- | -- | --
PA | 609 | Python Packaging Authority (PyPA) Governance | Dustin Ingram, Pradyun Gedam, Sumana Harihareswara
PA | 729 | Typing governance process | Jelle Zijlstra, Shantanu Jain
PA | 731 | C API Working Group Charter | Guido van Rossum, Petr Viktorin, Victor Stinner, Steve Dower, Irit Katriel
PA | 732 | The Python Documentation Editorial Board | Joanna Jablonski



https://peps.python.org/topic/governance/

miketheman commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+While there is overlap between the PyPA, Packaging-WG as well as the Python
+core team, the Steering Council is not well-placed to make decisions over
+Python packaging matters directly. Packaging only tangentially intersects with
+Python language evolution, packaging requires specialised expertise and deep
```
There's a part of me that wants the time machine to go back and make a very clear distinction between import packages and distribution packages.
Python's import package behavior is very much tied to the language's evolution, is it not?

Would it make sense to clarify the use of the term "packaging" early on to refer to distribution packaging in the context of this PEP to stave off further confusion?
Unless the intent is that the Packaging Council will actively work to evolve the Python language's changes to imports?

miketheman commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Whenever there is a vacancy during the regular council term, the council may
+vote to appoint a replacement to serve out the rest of the term.
+
+If a council member drops out of touch and cannot be contacted for a month or
```
Should the phrasing account for scheduled out of touch time periods? If someone is intentionally away for over a month and has notified the rest of the council, would this affect them?

```suggestion
If a council member unexpectedly drops out of touch and cannot be contacted for a month or
```

I understand that this is based on PEP 13, but that's not a reason to not improve/clarify if possible - and then update PEP 13 as well.

warsaw replied:
> If someone is intentionally away for over a month and has notified the rest of the council, would this affect them?

I would think not, but hopefully the member and the council can work together to discuss whether an extended, but in-touch absence should necessitate a resignation, in the best interests of the community.

> I understand that this is based on PEP 13, but that's not a reason to not improve/clarify if possible - and then update PEP 13 as well.

There's definitely a [process for that](https://peps.python.org/pep-0013/#changing-this-document), and there have been [3 amendments to date](https://peps.python.org/pep-0013/#history-of-amendments).

pradyunsg replied:
I'm inclined to say that this isn't an editorial concern but a concern with the specifics of the proposal -- and that we should discuss any such specifics in the corresponding d.p.o. thread.

I've taken note of this FWIW, and will make sure to flag this in that discussion.

JelleZijlstra commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Delegations
+-----------
+
+The Python Steering Council would be expected to delegate decision making to
```
```suggestion
The Python Steering Council is expected to delegate decision making to
```
And remove "would" in the rest of this paragraph too.

JelleZijlstra commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+Term
+----
+
+There shall be two cohorts of council members: Cohort A composed of two (2)
```
```suggestion
There shall be two cohorts of council members: Cohort A composed of two
```
And remove this later afterwards. Parenthesizing the digit doesn't remove any ambiguity, it just makes the text sound more legalese.

hugovk replied:
And could also replace some digits with words elsewhere ("2"->"two" and so on). A common rule is words for one to nine, digits for 10 and up.

pradyunsg commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,463 @@
+PEP: 772
```
My thinking with not using 8XXX numbering here was that all [PEPs under the Governance topic](https://peps.python.org/topic/governance/) are currently informational, with all process PEPs using regular PEP numbers.

warsaw replied:
It's kind of difficult to find the governance PEPs in the numerical listing.  Yes, there's the governance topic which makes it easier.  I do like that all the SC governance peps are somewhat closely numbered (and 13 is in the low reserves), especially since each annual election PEP just flows after the last 81xx.

I think the PC given its similarity to SC and especially with the expected voting PEPs warrants putting them in their own block, but I'll leave that final decision up to @pradyunsg

pradyunsg replied:
Let's discuss this out-of-band, since I realise what I did here isn't what we had agreed on in the pre-PEP discussions. :)

pradyunsg replied:
Referencing a relavant comment from outside this thread: https://github.com/python/peps/pull/4218#issuecomment-2611006165

pradyunsg commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+
+This PEP proposes a Python Packaging Council with broad authority over
+packaging standards, tools, and implementations. Like the Python Steering
+Council, they seek to exercise this authority as rarely as possible; instead,
```
(updated locally, will push)

pradyunsg commented on peps/pep-0772.rst:
```diff
@@ -0,0 +1,455 @@
+* look for ways to use their direct authority as little as possible – seeking
+  consensus/consent over voting.
+
+To use its powers, the council votes. Every council member must either vote or
+explicitly abstain. Members with conflicts of interest on a particular vote
+must abstain. Passing requires support from a majority of non-abstaining
+council members.
```
It does -- and I think this specific concern is covered by that.

If we need want to change procedures here to accomodate for something related to that, i.e. deviate further from PEP 13's model here, I suggest we discuss that in the discuss.python.org thread instead of on this PR.

pradyunsg commented:
Thanks everyone for the reviews here!

Given that there's over a hundred comments on this PR already... I have a couple of "housekeeping"-style requests:

- I'd appreciate if we avoid extended discussion _outside_ of inline review comments.
  - Replies outside of inline "review comments" are much easier to lose track of when the discussion gets really long with reviews from multiple individuals (which is what is going on here, right now).
  - I'm marking things as resolved to close out relevant subtopics in our discussions here.
- If I've marked one of your comments as resolved and haven't actually addressed it (and it wasn't a nit pick), please feel welcome mark it as unresolved and add a reply there!

There's still more stuff to be addressed from these reviews... I'll come around to those in my next burst of addressing review comments here (it's late today).

hugovk reviewed (approved):
I think my suggestions have been addressed. Thanks for working on this!

willingc reviewed (approved):
I'm good with where things stand right now. Thanks @warsaw and @pradyunsg for moving this forward.

warsaw reviewed (commented):
Thanks @willingc ; I'm going to get the last few things resolved and get this PEP published.
